File: 2nd Semester/Graphs/A3/graph.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
#include <set>
#include <map>

using namespace std;

class Graph
{
private:
  int v, e;
  map<int, vector<pair<int, int>>> in_edges;
  map<int, vector<pair<int, int>>> out_edges;
  map<int, int> cost;

public:
  Graph(int v = 0, int e = 0)
  {
    this->v = v;
    this->e = e;
  }
  ~Graph()
  {
  }
  void setNumberOfVertices(int v)
  {
    this->v = v;
  }
  void setNumberOfEdges(int e)
  {
    this->e = e;
  }
  int getNumberOfVertices()
  {
    return this->v;
  }
  int getNumberOfEdges()
  {
    return this->e;
  }
  pair<int, int> getEndpointsOfEdge(int edge)
  {
    for (auto it : out_edges)
      for (auto it2 : it.second)
        if (it2.second == edge)
          return make_pair(it.first, it2.first);
    return make_pair(-1, -1);
  }
  vector<int> getVerticesOfgraph()
  {
    vector<int> vertices;
    for (auto it : out_edges)
      vertices.push_back(it.first);
    return vertices;
  }
  int getIdOfEdge(int u, int v)
  {
    for (auto it : out_edges[u])
      if (it.first == v)
        return it.second;
    return -1;
  }
  bool checkIfEdgeExists(int u, int v)
  {
    return this->getIdOfEdge(u, v) != -1;
  }
  int getInDegreeOfVertex(int v)
  {
    return in_edges[v].size();
  }
  int getOutDegreeOfVertex(int v)
  {
    return out_edges[v].size();
  }
  vector<pair<int, int>> getInEdgesOfVertex(int v)
  {
    return in_edges[v];
  }
  vector<pair<int, int>> getOutEdgesOfVertex(int v)
  {
    return out_edges[v];
  }
  void setCostOfEdge(int u, int v, int cost)
  {
    if (this->checkIfEdgeExists(u, v))
    {
      int id = this->getIdOfEdge(u, v);
      this->cost[id] = cost;
    }
  }
  int getCostOfEdge(int u, int v)
  {
    if (this->checkIfEdgeExists(u, v))
    {
      int id = this->getIdOfEdge(u, v);
      return this->cost[id];
    }
    return -1;
  }
  Graph getCopyOfGraph()
  {
    Graph g(this->v, this->e);
    g.in_edges = this->in_edges;
    g.out_edges = this->out_edges;
    g.cost = this->cost;
    return g;
  }
  void addEdge(int u, int v, int cost)
  {
    if (this->checkIfEdgeExists(u, v))
      return;
    int id = this->e;
    ++this->e;
    this->out_edges[u].push_back(make_pair(v, id));
    this->in_edges[v].push_back(make_pair(u, id));
    this->cost[id] = cost;
  }
  void removeEdge(int u, int v)
  {
    if (!this->checkIfEdgeExists(u, v))
      return;
    int id = this->getIdOfEdge(u, v);
    this->out_edges[u].erase(remove_if(this->out_edges[u].begin(), this->out_edges[u].end(), [id](pair<int, int> p)
                                       { return p.second == id; }),
                             this->out_edges[u].end());
    this->in_edges[v].erase(remove_if(this->in_edges[v].begin(), this->in_edges[v].end(), [id](pair<int, int> p)
                                      { return p.second == id; }),
                            this->in_edges[v].end());
    this->cost.erase(id);
    --this->e;
  }
  void addVertex(int v)
  {
    if (this->out_edges.find(v) != this->out_edges.end())
      return;
    ++this->v;
    this->out_edges[v] = vector<pair<int, int>>();
    this->in_edges[v] = vector<pair<int, int>>();
  }
  void removeVertex(int v)
  {
    if (this->out_edges.find(v) == this->out_edges.end())
      return;
    for (auto it : this->out_edges[v])
      this->removeEdge(v, it.first);
    for (auto it : this->in_edges[v])
      this->removeEdge(it.first, v);
    this->out_edges.erase(v);
    this->in_edges.erase(v);
    --this->v;
  }
  void generateRandomGraph(int v, int e)
  {
    this->v = v;
    this->e = e;
    for (int i = 0; i < v; ++i)
      this->addVertex(i);
    for (int i = 0; i < e; ++i)
    {
      int u = rand() % v;
      int v = rand() % v;
      int cost = rand() % 100;
      this->addEdge(u, v, cost);
    }
  }
};

ifstream fin("graph1k.in");
ofstream fout("graph1k.out");

class UI
{
private:
  Graph g;

public:
  UI()
  {
  }
  ~UI()
  {
  }
  void printTheNumberOfVertices()
  {
    cout << "The Number of Vertices is " << g.getNumberOfVertices() << '\n';
  }
  void printTheNumberOfEdges()
  {
    cout << "The Number of Edges is " << g.getNumberOfEdges() << "\n";
  }
  void printTheVerticesOfGraphOnScreen()
  {
    cout << "The Vertices of the Graph are: ";
    vector<int> vertices = g.getVerticesOfgraph();
    for (auto it : vertices)
      cout << it << " ";
    cout << endl;
  }
  void printTheEdgesOfGraphOnScreen()
  {
    cout << "The Edges of the Graph are: " << endl;
    vector<int> vertices = g.getVerticesOfgraph();
    for (auto it : vertices)
    {
      vector<pair<int, int>> edges = g.getOutEdgesOfVertex(it);
      for (auto it2 : edges)
        cout << it << " " << it2.first << " " << g.getCostOfEdge(it, it2.first) << endl;
    }
  }
  void readGraphFromFile()
  {
    int v, e;
    fin >> v >> e;
    for (int i = 0; i < v; ++i)
      g.addVertex(i);
    for (int i = 0; i < e; ++i)
    {
      int u, v, cost;
      fin >> u >> v >> cost;
      g.addEdge(u, v, cost);
    }
    cout << "Graph read from file successfully! " << endl;
  }
  void writeGraphToFile()
  {
    fout << g.getNumberOfVertices() << " " << g.getNumberOfEdges() << endl;
    vector<int> vertices = g.getVerticesOfgraph();
    for (auto it : vertices)
    {
      vector<pair<int, int>> edges = g.getOutEdgesOfVertex(it);
      for (auto it2 : edges)
        fout << it << " " << it2.first << " " << g.getCostOfEdge(it, it2.first) << endl;
    }
    cout << "Graph written to file successfully" << endl;
  }
  void printIdOfEdge()
  {
    int u, v;
    cout << "Enter the two vertices of the edge" << endl;
    cin >> u >> v;
    int idOfEdge = g.getIdOfEdge(u, v);
    if (idOfEdge == -1)
      cout << "Edge does not exist" << endl;
    else
      cout << "The id of the edge is " << idOfEdge << endl;
  }
  void printInDegreeAndOutDegreeOfVertex()
  {
    cout << "Which vertex do you want to check?" << endl;
    int v;
    cin >> v;
    cout << "In degree of vertex " << v << " is " << g.getInDegreeOfVertex(v) << endl;
    cout << "Out degree of vertex " << v << " is " << g.getOutDegreeOfVertex(v) << endl;
  }
  void printTheOutboundEdgesOfVertex()
  {
    cout << "Which vertex do you want to check?" << endl;
    int v;
    cin >> v;
    vector<pair<int, int>> edges = g.getOutEdgesOfVertex(v);
    if (edges.size() == 0)
    {
      cout << "There are no outbound edges to the vertex!" << endl;
      return;
    }
    cout << "The outbound edges of vertex " << v << " are: " << endl;
    for (auto it : edges)
      cout << v << " " << it.first << " " << g.getCostOfEdge(v, it.first) << endl;
  }
  void printTheInboundEdgesOfVertex()
  {
    cout << "Which vertex do you want to check?" << endl;
    int v;
    cin >> v;
    vector<pair<int, int>> edges = g.getInEdgesOfVertex(v);
    if (edges.size() == 0)
    {
      cout << "There are no inbound edges to the vertex!" << endl;
      return;
    }
    cout << "The inbound edges of vertex " << v << " are: " << endl;
    for (auto it : edges)
      cout << it.first << " " << v << " " << g.getCostOfEdge(it.first, v) << endl;
  }
  void printTheCostOfEdge()
  {
    cout << "Enter the two vertices of the edge" << endl;
    int u, v;
    cin >> u >> v;
    int cost = g.getCostOfEdge(u, v);
    if (cost == -1)
      cout << "Edge does not exist" << endl;
    else
      cout << "The cost of the edge is " << cost << endl;
  }
  void printTheEndpointsOfAnEdge()
  {
    cout << "Enter the id of the edge" << endl;
    int id;
    cin >> id;
    pair<int, int> endpoints = g.getEndpointsOfEdge(id);
    if (endpoints.first == -1)
      cout << "Edge does not exist" << endl;
    else
      cout << "The endpoints of the edge are " << endpoints.first << " " << endpoints.second << endl;
  }
  void addEdge()
  {
    cout << "Enter the two vertices of the edge and the cost" << endl;
    int u, v, cost;
    cin >> u >> v >> cost;
    g.addEdge(u, v, cost);
  }
  void removeEdge()
  {
    cout << "Enter the two vertices of the edge" << endl;
    int u, v;
    cin >> u >> v;
    g.removeEdge(u, v);
  }
  void addVertex()
  {
    cout << "Enter the id of the vertex" << endl;
    int v, n = g.getNumberOfVertices();
    cin >> v;
    if (n == 0)
    {
      cout << "The graph is empty. You have to set a dimension first" << endl;
      int n;
      cin >> n;
      g.setNumberOfVertices(n);
    }
    if (v < 0 || v >= n + 1)
    {
      cout << "Invalid vertex" << endl;
      return;
    }
    g.addVertex(v);
  }
  void removeVertex()
  {
    cout << "Enter the id of the vertex" << endl;
    int v, n = g.getNumberOfVertices();
    cin >> v;
    if (n == 0)
    {
      cout << "The graph is empty. There is nothing to remove" << endl;
      return;
    }
    if (v < 0 || v >= n)
    {
      cout << "Invalid vertex" << endl;
      return;
    }
    g.removeVertex(v);
  }
  void printMenuOfOptions()
  {
    cout << "1. Print the number of vertices" << endl;
    cout << "2. Print the number of edges" << endl;
    cout << "3. Print the vertices of the graph" << endl;
    cout << "4. Print the edges of the graph" << endl;
    cout << "5. Read the graph from a file" << endl;
    cout << "6. Write the graph to a file" << endl;
    cout << "7. Print the id of an edge" << endl;
    cout << "8. Print the in degree and out degree of a vertex" << endl;
    cout << "9. Print the outbound edges of a vertex" << endl;
    cout << "10. Print the inbound edges of a vertex" << endl;
    cout << "11. Print the cost of an edge" << endl;
    cout << "12. Print the endpoints of an edge" << endl;
    cout << "13. Add an edge" << endl;
    cout << "14. Remove an edge" << endl;
    cout << "15. Add a vertex" << endl;
    cout << "16. Remove a vertex" << endl;
    cout << "17. Exit" << endl;
  }
};
int main()
{
  UI ui;
  while (1)
  {
    ui.printMenuOfOptions();
    int option;
    cin >> option;
    switch (option)
    {
    case 1:
      ui.printTheNumberOfVertices();
      break;
    case 2:
      ui.printTheNumberOfEdges();
      break;
    case 3:
      ui.printTheVerticesOfGraphOnScreen();
      break;
    case 4:
      ui.printTheEdgesOfGraphOnScreen();
      break;
    case 5:
      ui.readGraphFromFile();
      break;
    case 6:
      ui.writeGraphToFile();
      break;
    case 7:
      ui.printIdOfEdge();
      break;
    case 8:
      ui.printInDegreeAndOutDegreeOfVertex();
      break;
    case 9:
      ui.printTheOutboundEdgesOfVertex();
      break;
    case 10:
      ui.printTheInboundEdgesOfVertex();
      break;
    case 11:
      ui.printTheCostOfEdge();
      break;
    case 12:
      ui.printTheEndpointsOfAnEdge();
      break;
    case 13:
      ui.addEdge();
      break;
    case 14:
      ui.removeEdge();
      break;
    case 15:
      ui.addVertex();
      break;
    case 16:
      ui.removeVertex();
      break;
    case 17:
      cout << "Goodbye!" << endl;
      return 0;
    default:
      cout << "Invalid option" << endl;
    }
  }
  return 0;
}

File: 2nd Semester/Graphs/A3/graph.py
import copy


class Graph:
    def __init__(self, nr_of_vertices=0):
        """
        :param nr_of_vertices: the number of vertices in the graph. Default value is 0

        Default constructor of our graph. It initializes the following fields:

        self.__nr_of_vertices: the number of vertices in the graph

        self.__nr_of_edges: the number of edges in the graph

        self.__in_edges: a dictionary that stores the inbound edges of a vertex

        self.__out_edges: a dictionary that stores the outbound edges of a vertex

        self.__edges_cost: a dictionary that stores the cost of an edge
         """
        self.__nr_of_vertices = nr_of_vertices
        self.__nr_of_edges = 0
        self.__in_edges = {}
        self.__out_edges = {}
        self.__edges_cost = {}

    """Setters and getters for our functions"""

    def set_number_of_vertices(self, nr_of_vertices):
        """
        :param nr_of_vertices: the number of vertices with which our graph will be updated

        Set value for the number of vertices in the graph
        """
        self.__nr_of_vertices = nr_of_vertices

    def get_number_of_vertices(self):
        """:return: The number of vertices of our graph"""
        return self.__nr_of_vertices

    def get_set_of_vertices(self):
        """
        :return: 
        The set of vertices of our graph"""
        return set(self.__in_edges.keys())

    def set_number_of_edges(self, nr_of_edges):
        """
        :param nr_of_edges: the number of edges with which our graph will be updated

        Set value for the number of edges in the graph
        """
        self.__nr_of_edges = nr_of_edges

    def get_endpoints_of_edge(self, edge_id):
        """
        :param edge_id: the id of the edge for which we want to retrieve the endpoints

        We iterate through the outbound edges of each vertex and check if the edge_id is the same as the one we are looking for.
        ...

        :return: a tuple (x, y) where x is the start node and y is the end node of the edge, or (-1, -1) if the edge does not exist

        """
        for x in self.__out_edges:
            for y in self.__out_edges[x]:
                if self.__out_edges[x][y] == edge_id:
                    return x, y
        return -1, -1

    def return_id_of_edge(self, x, y):
        """
        :param x: the start node of the edge
        :param y: the end node of the edge
        :return: the id of the edge (x, y) if it exists, -1 otherwise
        """
        if x in self.__out_edges and y in self.__out_edges[x]:
            return self.__out_edges[x][y]
        return -1

    def get_number_of_edges(self):
        """
        :return:
        The number of edges of our graph"""
        return self.__nr_of_edges

    def get_out_edges(self):
        """
        :return:
        The outbound edges of our graph"""
        return dict(sorted(self.__out_edges.items()))

    def get_in_edges(self):
        """
        :return:
        The inbound edges of our graph"""
        return dict(sorted(self.__in_edges.items()))

    def get_in_degree(self, v):
        """
        :param v: the vertex for which we want to know the in degree
        :return: The in degree of a vertex v
        """
        if v in self.__in_edges:
            return len(self.__in_edges[v])

    def get_out_degree(self, v):
        """
        :param v: the vertex for which we want to know the out degree

        :return: The out degree of a vertex v
        """
        if v in self.__out_edges:
            return len(self.__out_edges[v])

    def get_outbound_neighbours(self, v):
        """
        :param v: the vertex for which we want to retrieve the outbound neighbours

        :return:
        Since we store the inbound neighbours of a vertex in a dictionary, we simply have to return the set of keys
        """
        return set(self.__out_edges[v].keys())

    def get_inbound_neighbours(self, v):
        """
        :param v: the vertex for which we want to retrieve the inbound neighbours

        :return:
        Since we store the inbound neighbours of a vertex in a dictionary, we simply have to return the set of keys
        """
        return set(self.__in_edges[v].keys())

    def get_cost_of_edge(self, edge_id):
        """
        :param edge_id: the id of the edge for which we want to retrieve the cost

        :return: The cost of the edge with the given id
        """
        return self.__edges_cost[edge_id]

    def modify_cost_of_edge(self, edge_id, cost):
        """
        :param edge_id: the id of the edge for which we want to set the cost
        :param cost: the cost of the edge
        """
        self.__edges_cost[edge_id] = cost

    def check_if_edge_exists(self, x, y):
        """
        :param x: the start node of the edge
        :param y: the end node of the edge
        :return: True if the edge (x, y) exists, False otherwise

        In order to check if an edge exists, we simply call the function return_id_of_edge and check if it returns -1 or not
        """
        return self.return_id_of_edge(x, y) != -1

    def add_vertex_in_graph(self, v):
        """
        :param v: the vertex we have to add to the grap
        """
        if not v in self.__in_edges:
            self.__in_edges[v] = {}
            self.__out_edges[v] = {}

    def remove_vertex_from_graph(self, v):
        """
        :param v: the vertex we have to remove from the graph
        """
        if v in self.__in_edges:
            self.__nr_of_vertices -= 1
            self.__nr_of_edges -= len(self.__in_edges[v]) + \
                len(self.__out_edges[v])
            for x in self.__in_edges[v]:
                del self.__edges_cost[self.__in_edges[v][x]]
                del self.__out_edges[x][v]
            for x in self.__out_edges[v]:
                del self.__edges_cost[self.__out_edges[v][x]]
                del self.__in_edges[x][v]
            del self.__in_edges[v]
            del self.__out_edges[v]

    def add_edge_to_graph(self, start_node, end_node, cost):
        """
        :param start_node: the start node of the edge
        :param end_node: the end node of the edge
        :param edge_id: the id of the edge
        """
        self.add_vertex_in_graph(start_node)
        self.add_vertex_in_graph(end_node)
        self.__in_edges[end_node][start_node] = self.__nr_of_edges
        self.__out_edges[start_node][end_node] = self.__nr_of_edges
        self.__edges_cost[self.__out_edges[start_node][end_node]] = cost
        self.__nr_of_edges += 1

    def remove_edge_from_graph(self, start_node, end_node):
        """
        :param start_node: the start node of the edge
        :param end_node: the end node of the edge
        """
        if self.check_if_edge_exists(start_node, end_node):
            del self.__edges_cost[self.__out_edges[start_node][end_node]]
            del self.__out_edges[start_node][end_node]
            del self.__in_edges[end_node][start_node]
            self.__nr_of_edges -= 1

    def get_copy_of_graph(self):
        """
        Function that returns a copy of the graph

        :return: a copy of the graph
        """
        g = Graph(self.__nr_of_vertices)
        g.__in_edges = copy.deepcopy(self.__in_edges)
        g.__out_edges = copy.deepcopy(self.__out_edges)
        g.__edges_cost = copy.deepcopy(self.__edges_cost)
        return g

    def set_cost_on_edge(self, edge_id, cost):
        """
        :param edge_id: the id of the edge for which we want to set the cost
        :param cost: the cost of the edge
        """
        self.__edges_cost[edge_id] = cost
